File: framework/src/play-caffeine-cache/src/main/java/play/cache/caffeine/CaffeineCacheManager.java
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package play.cache.caffeine;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import com.github.benmanes.caffeine.cache.Caffeine;
import com.typesafe.config.Config;

public class CaffeineCacheManager {
    private final ConcurrentMap<String, NamedCaffeineCache> cacheMap = new ConcurrentHashMap<>(16);
    private Config config;

    public CaffeineCacheManager(Config config) {
        this.config = config;
    }

    @SuppressWarnings("unchecked")
    public <K, V> NamedCaffeineCache<K, V> getCache(String cacheName) {
        NamedCaffeineCache<K, V> namedCache = cacheMap.getOrDefault(cacheName, null);

        // if the cache is null we have to create it
        if (namedCache == null) {
            Caffeine cacheBuilder = getCacheBuilder(cacheName);

            namedCache = new NamedCaffeineCache(cacheName, cacheBuilder.build());

            cacheMap.put(cacheName, namedCache);
        }

        return namedCache;
    }

    private Caffeine getCacheBuilder(String cacheName) {
        Caffeine cacheBuilder;
        CaffeineDefaultExpiry defaultExpiry = new CaffeineDefaultExpiry();
        Config caches = config.getConfig("caches");
        Config defaults = config.getConfig("defaults");
        Config cacheConfig;

        if (caches.hasPath(cacheName)) {
            cacheConfig = caches.getConfig(cacheName).withFallback(defaults);
        } else {
            cacheConfig = defaults;
        }

        cacheBuilder = CaffeineParser.from(cacheConfig).expireAfter(defaultExpiry);

        return cacheBuilder;
    }
}

File: framework/src/play-caffeine-cache/src/main/scala/play/api/cache/caffeine/CaffeineCacheManager.scala
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.cache.caffeine

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap

import com.github.benmanes.caffeine.cache.Caffeine
import com.typesafe.config.Config
import play.cache.caffeine.{ CaffeineDefaultExpiry, CaffeineParser, NamedCaffeineCache }

import com.github.benmanes.caffeine.cache.Cache

class CaffeineCacheManager(private var config: Config) {

  private val cacheMap: ConcurrentMap[String, NamedCaffeineCache[_, _]] =
    new ConcurrentHashMap(16)

  def getCache[K, V](cacheName: String): NamedCaffeineCache[K, V] = {
    var namedCache: NamedCaffeineCache[K, V] =
      cacheMap.getOrDefault(cacheName, null).asInstanceOf[NamedCaffeineCache[K, V]]
    // if the cache is null we have to create it

    if (namedCache == null) {
      val cacheBuilder: Caffeine[K, V] = getCacheBuilder(cacheName).asInstanceOf[Caffeine[K, V]]
      namedCache = new NamedCaffeineCache[K, V](cacheName, cacheBuilder.build().asInstanceOf[Cache[K, V]])
      cacheMap.put(cacheName, namedCache.asInstanceOf[NamedCaffeineCache[_, _]])
    }
    namedCache
  }

  private[caffeine] def getCacheBuilder(cacheName: String): Caffeine[_, _] = {
    var cacheBuilder: Caffeine[_, _] = null
    val defaultExpiry: CaffeineDefaultExpiry = new CaffeineDefaultExpiry()
    val caches: Config = config.getConfig("caches")
    val defaults: Config = config.getConfig("defaults")
    var cacheConfig: Config = null
    cacheConfig =
      if (caches.hasPath(cacheName))
        caches.getConfig(cacheName).withFallback(defaults)
      else defaults
    cacheBuilder = CaffeineParser.from(cacheConfig).expireAfter(defaultExpiry)
    cacheBuilder
  }

}
